### PR TITLE
finish schema_inferer

### PR DESCRIPTION
**Pull request by zhexuany**
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tikv-client-java/37)

### CODE REVIEW

**Comment by zhexuany** [comment]
@ilovesoup  only for review. Not ready yet, still working on it.

File: src/main/java/com/pingcap/tikv/RegionStoreClient.java [hunk 1]
```diff
@@ -212,24 +213,14 @@ public void close() throws Exception {
 
     public static RegionStoreClient create(Region region, Store store, TiSession session) {
         RegionStoreClient client = null;
-        try {
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/RegionStoreClient.java]: what's the purpose of this change?

File: src/main/java/com/pingcap/tikv/SelectBuilder.java [hunk 2]
```diff
@@ -54,31 +43,31 @@ public static SelectBuilder newBuilder(Snapshot snapshot, TiTableInfo table) {
...
 
   public SelectBuilder setTimeZoneOffset(long offset) {
-    builder.setTimeZoneOffset(offset);
+//    builder.setTimeZoneOffset(offset);
     return this;
   }
 
   public SelectBuilder setTimestamp(long timestamp) {
-    builder.setStartTs(timestamp);
+//    builder.setStartTs(timestamp);
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/SelectBuilder.java]: why is this?

File: src/main/java/com/pingcap/tikv/codec/DecimalUtils.java [hunk 3]
```diff
@@ -30,18 +30,26 @@
      * @param cdi cdi is source data.
      * */
     public static double readDecimalFully(CodecDataInput cdi) {
-        if (cdi.size() < 3) {
+        if (cdi.available() < 3) {
             throw new IllegalArgumentException("insufficient bytes to read value");
         }
+
+        MyDecimal dec = new MyDecimal();
+        List<Integer> data = new ArrayList<>();
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/codec/DecimalUtils.java]: Move original conversation here from other PR.
"I think I already did that way. You could scroll down a little bit and you would find out that Ints.toArray."

If you use a data arraylist<Integer>, all things pushed inside is boxed into Integer which is an object.
Boxing is usually not a problem. But it might be too much for a decoding utility and Ints.toArray does not reduce such costs. 

You could dig a little bit deeper before reply and you would find what comment really means. :)

https://docs.oracle.com/javase/8/docs/technotes/guides/language/autoboxing.html

You might try guava Ints.ensureCapacity or use trove's IntArrayList (essentially the same). 

I did a very simple test for trove:
https://gist.github.com/ilovesoup/c9aa2f1c5d51229c4dd160b399488c21

And the test result on my mac is:
TIntArrayList:0.010688442
ArrayList:0.549776375
TIntArrayList:0.04866792
ArrayList:0.745774716
TIntArrayList:0.00415313
ArrayList:0.47087427
TIntArrayList:0.005395041
ArrayList:0.473152197
TIntArrayList:0.004185768
ArrayList:0.432565985

**Review by ilovesoup — COMMENTED** [review]
I didn't bother moving comments from the closed PR. Let's discuss in person tomorrow.

File: src/main/java/com/pingcap/tikv/RegionStoreClient.java (hunk 1 above)
Comment (zhexuany) [reply src/main/java/com/pingcap/tikv/RegionStoreClient.java]: Ah, Intellij says this way is better. lol.

File: src/main/java/com/pingcap/tikv/SelectBuilder.java (hunk 2 above)
Comment (zhexuany) [reply src/main/java/com/pingcap/tikv/SelectBuilder.java]: Code have not been cleaned up yet.  Working on it right now.

File: src/main/java/com/pingcap/tikv/codec/DecimalUtils.java (hunk 3 above)
Comment (zhexuany) [reply src/main/java/com/pingcap/tikv/codec/DecimalUtils.java]: AutoBoxing is known as very heavy. I got this point. I was to rush to jump to the conclusion and will do it right now.  You are absolutely right. `Ints.toArray`could not reduce  the burden. Another approbate approach is needed here. Thanks for the very detail investigation.

File: src/main/java/com/pingcap/tikv/operation/SchemaInfer.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        private boolean isSingledGroup;
+        private boolean isSimpleSelect;
+        private List<FieldType> fieldTypes;
+        private List<FieldType> finalFieldTypes;
+        private TiSelectRequest tiSelectRequest;
+
+        private TiFieldType(TiSelectRequest tiSelectRequest) {
+            fieldTypes = new ArrayList<>();
+            finalFieldTypes = new ArrayList<>();
+             if (tiSelectRequest.getGroupBys().size() == 0) {
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/operation/SchemaInfer.java]: indentation

File: src/main/java/com/pingcap/tikv/operation/SchemaInfer.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+ * the same type or not. If yes, no need for casting. If no, casting is needed here.
+ */
+public class SchemaInfer {
+
+    public static TiFieldType create(TiSelectRequest tiSelectRequest) {
+        return new TiFieldType(tiSelectRequest);
+    }
+
+    @Data
+    public static class TiFieldType {
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/operation/SchemaInfer.java]: Naming is not intuitive. Shouldn't be TiCoprocessorRequest or something similar?

**Review by ilovesoup — COMMENTED** [review]
After comment addressed, LGTM.

**Comment by zhexuany** [comment]
SchemaInfer's code is too ugly. More refactoring on the way.

**Review by ilovesoup — APPROVED** [review]
LGTM

**Comment by ilovesoup** [comment]
Column ref for group by is ok but need expression support for group by in coprocessor
